# --- T2-COPYRIGHT-BEGIN ---
# t2/package/*/chromium/hotfix-pre-clang23.patch
# Copyright (C) 2026 The T2 SDE Project
# SPDX-License-Identifier: GPL-2.0 or patched project license
# --- T2-COPYRIGHT-END ---

--- chromium-151.0.7922.75/build/config/sanitizers/sanitizers.gni.vanilla
+++ chromium-151.0.7922.75/build/config/sanitizers/sanitizers.gni
@@ -534,14 +534,19 @@
       cflags = [
         "-fsanitize=${invoker.sanitizer}",
         "-fsanitize-trap=${invoker.sanitizer}",
+      ]
 
+      # T2: clang only knows this flag from 23 on - see the
+      # hotfix-pre-clang23.patch header.
+      if (!t2_pre_clang23) {
         # Prevents `__has_feature(undefined_behavior_sanitizer)`
         # from evaluating true. Configs defined here are intended to
         # be usable even in release builds, i.e. as widely as possible.
         # It's important not to have full-on UBSan workarounds activate
         # just because we built support for a specific sanitizer.
-        "-fsanitize-ignore-for-ubsan-feature=${invoker.sanitizer}",
-      ]
+        cflags += [ "-fsanitize-ignore-for-ubsan-feature=${invoker.sanitizer}" ]
+      }
+
       if (defined(invoker.cflags)) {
         cflags += invoker.cflags
       }
--- chromium-151.0.7922.75/build/config/compiler/BUILD.gn.vanilla
+++ chromium-151.0.7922.75/build/config/compiler/BUILD.gn
@@ -578,7 +578,11 @@
     cflags += [ "-fcolor-diagnostics" ]
     if (!is_wasm) {
       if (!is_win) {
-        cflags += [ "-fdiagnostics-show-inlining-chain" ]
+        # T2: clang only knows this flag from 23 on - see the
+        # hotfix-pre-clang23.patch header.
+        if (!t2_pre_clang23) {
+          cflags += [ "-fdiagnostics-show-inlining-chain" ]
+        }
       } else {
         # Combine after https://github.com/llvm/llvm-project/pull/192241
         cflags += [ "/clang:-fdiagnostics-show-inlining-chain" ]
@@ -617,7 +621,9 @@
     # The performance improvement does not seem worth the risk. See
     # https://crbug.com/484082200 for background and https://crrev.com/c/7593035
     # for discussion.
-    if (!is_wasm) {
+    # T2: clang only knows -fno-lifetime-dse from 23 on - see the
+    # hotfix-pre-clang23.patch header.
+    if (!is_wasm && !t2_pre_clang23) {
       cflags += [ "-fno-lifetime-dse" ]
     }
 
@@ -1584,7 +1590,13 @@
   # the related warning about "unsafe-buffer-usage-in-static-sized-array",
   # since we know that the array bounds sanitizing will catch any out-of-
   # bounds accesses.
-  cflags = [ "-Wno-unsafe-buffer-usage-in-static-sized-array" ]
+  #
+  # T2: the warning itself only exists from clang 23 on, so suppressing it
+  # earlier just yields -Wunknown-warning-option on every TU - see the
+  # hotfix-pre-clang23.patch header.
+  if (!t2_pre_clang23) {
+    cflags = [ "-Wno-unsafe-buffer-usage-in-static-sized-array" ]
+  }
 }
 
 # Enable UBSan subset to ensure we always hit `return` statement in non-void
@@ -1933,10 +1945,14 @@
       cflags += [
         # TODO(crbug.com/432275627): Fix and re-enable.
         "-Wno-uninitialized-const-pointer",
+      ]
 
+      # T2: the warning itself only exists from clang 23 on - see the
+      # hotfix-pre-clang23.patch header.
+      if (!t2_pre_clang23) {
         # TODO(crbug.com/495753203): Fix and re-enable.
-        "-Wno-unused-but-set-global",
-      ]
+        cflags += [ "-Wno-unused-but-set-global" ]
+      }
     }
 
     cflags_cc += [
